[nexus] Add stubs for image APIs

smklein opened the pull request:
First part of #734

smklein commented on nexus/src/external_api/params.rs:
```diff
@@ -229,6 +229,19 @@ pub struct DiskIdentifier {
+pub struct ImageCreate {
+    /// common identifying metadata
+    #[serde(flatten)]
+    pub identity: IdentityMetadataCreateParams,
+}
```
We need more here to actually be able to create the image.

Do we want this to be an enum implying "Please provide one of a {file, snapshot UUID, URL}"? Do we want that to all be done through the same endpoint?

smklein replied:
Added an enum of "snapshot UUID or URL". We can continue to add extra params as necessary.

smklein commented on common/src/sql/dbinit.sql:
```diff
@@ -370,6 +370,26 @@ CREATE INDEX ON omicron.public.disk (
+    time_deleted TIMESTAMPTZ,
+
+    project_id UUID,
+    volume_id UUID NOT NULL,
```
Once the image has been constructed, it'll refer to a particular volume.

For the case of e.g. POST-ing an image by URL, would we immediately download + store the image as a part of invoking the POST request?

- If yes: Cool, this structure should work. We'll create a saga that does that work.
- If no: This structure may need to be more complex - we'll need to store the URL too.

jmpesp replied:
This is currently an unsolved problem, though that answer will likely be the same as others: some block migration task will move blocks from the read only parent to the volume region set(s), then remove the read only parent later. We can't gate the instance boot or image creation on that download though.

smklein replied:
Okay. I'll leave this question open as an implementation detail for follow-up PRs

smklein commented on common/src/sql/dbinit.sql:
```diff
@@ -370,6 +370,26 @@ CREATE INDEX ON omicron.public.disk (
 ) WHERE
     time_deleted IS NULL AND attach_instance_id IS NOT NULL;
 
+CREATE TABLE omicron.public.image (
```
RFD 4 mentions the following fields which are not yet stored here (nor a part of the API):
- version
- digest
- status

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+ * Images
+ */
+
+/// List global images.
```
```suggestion
/** List global images
 * 
 * Returns a list of all the global images. Global images are returned sorted by creation date, with the most recent images appearing first.
 */
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Create a global image.
```
```suggestion
/** Create a global image
 *
 * Create a new global image.
 */
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+    image_name: Name,
+}
+
+/// Get a global image.
```
```suggestion
/** Get a global image
 * 
 * Returns the details of a specific global image.
 */
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Delete a global image.
```
```suggestion
/** Delete a global image
 * 
 * Permanently delete a global image. This operation cannot be undone. Any instances using the global image will continue to run, however new instances can not be created with this image.
 */
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// List images in a project.
```
```suggestion
/** List images
 *
 * List images in a project. The images are returned sorted by creation date, with the most recent images appearing first.
 */
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+#[endpoint {
+    method = GET,
+    path = "/organizations/{organization_name}/projects/{project_name}/images",
+    tags = ["images"],
```
```suggestion
    tags = ["projects"],
```

smklein replied:
Is there documentation on what tag should be associated with what endpoint? I think I might not be understanding this fully. I thought that an operation acting on:

```
path = "/organizations/{organization_name}/projects"
```

Would be tagged with `projects`, but an operation acting on:

```
path = "/organizations/{organization_name}/projects/{project_name}/foobar"
```

Would be tagged with `foobar(s)`.

Are *both* supposed to be tagged with `projects`? If we keep nesting, is it just `projects` all the way down?

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Create an instance in a project
```
```suggestion
/** Create an image
 *
 * Create a new image in a project.
 *./
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+#[endpoint {
+    method = POST,
+    path = "/organizations/{organization_name}/projects/{project_name}/images",
+    tags = ["images"]
```
```suggestion
    tags = ["projects"]
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+    image_name: Name,
+}
+
+/// Get an image in a project.
```
```suggestion
/** Get an image
 *
 * Get the details of a specific image in a project.
 */
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+#[endpoint {
+    method = GET,
+    path = "/organizations/{organization_name}/projects/{project_name}/images/{image_name}",
+    tags = ["images"],
```
```suggestion
    tags = ["projects"],
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+/// Delete an image from a project.
```
```suggestion
/** Delete an image
 *
 * Permanently delete an image from a project. This operation cannot be undone. Any instances in the project using the image will continue to run, however new instances can not be created with this image.
 */
```

jessfraz commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -1106,6 +1120,249 @@ async fn instance_disks_detach(
+#[endpoint {
+    method = DELETE,
+    path = "/organizations/{organization_name}/projects/{project_name}/images/{image_name}",
+    tags = ["images"],
```
```suggestion
    tags = ["projects"],
```

jessfraz commented:
sorry if the fmt is off in my suggestions. Wrote them by hand but will align well with #773 when merged

jessfraz commented:
Yeah that’s fair I missed 770, is there a way to grep replace of I can do by hand, the content is mostly what matters

smklein commented:
Hey @jessfraz , I incorporated your recommended comments. Regarding the tags, I added documentation for "images" to `tag-config.json` - but lemme know if you still think I'm using the wrong tags!

jessfraz commented:
Awesome, tags and docs look great!
